Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -53,10 +54,18 @@ extension MongoCollection {
         let reply = Document()
         var error = bson_error_t()
         if !mongoc_collection_insert_many(
+             // TODO SWIFT-139: include writeErrors and writeConcernErrors from reply in the error
             self._collection, &docPointers, documents.count, opts?.data, reply.data, &error) {
             throw MongoError.commandError(message: toErrorString(error))
         }
-        return InsertManyResult(insertedIds: documents.map { $0["_id"] })
+
+        // if we succeeded, then we should store + return every `_id`
```

Comment:
Interesting. Since [`mongoc_collection_insert_many()`](http://mongoc.org/libmongoc/current/mongoc_collection_insert_many.html) doesn't report `insertedIds`, we might actually have to interpret the errors to determine how to (a) construct the map or (b) redact failed inserts from the map (assuming we construct it before executing the insert).

I know that PHPLIB reports inserted IDs in its InsertManyResult class. Looking at [the source](https://github.com/mongodb/mongo-php-library/blob/1.4.2/src/Operation/InsertMany.php#L142), it appears we don't report the map if an exception is thrown. In that case, we let the exception from [`executeBulkWrite()`](http://php.net/manual/en/mongodb-driver-manager.executebulkwrite.php) propagate. That exception will include a PHPC [WriteResult](http://php.net/manual/en/class.mongodb-driver-writeresult.php), which is derived from libmongoc's reply document and does not include a map of inserted IDs.

Given the complexity of building this map after encountering an error (mentioned in the first paragraph of this comment), I'm not sure it's worth the trouble. I might suggest poking around other drivers to see how they handle this before expending the effort to implement the TODO item above.

Reply:
yes, building it does seem like it will be difficult, so maybe in the case of an error in `insertMany` or `bulkWrite` we should leave `insertedIds` out... either way, we can propagate the writeError / writeConcern error fields mentioned in the to-do.